### DISCUSSION TITLE
Sending Fw::Buffer to Drv::Udp

### DISCUSSION THREAD

**Discussion by chroco** [nasa/fprime#3079] [Q&A] [answered]
I'm new to Fprime and I'm learning how to use ports. We have a rust driver for our radio that opens a UDP socket that accepts EDL packets and then transmits them. 

For starters, I just want to be able to test this functionality within fprime using Drv::Udp with server and client threads. I am trying to properly wire the ports to use a FW::Buffer so that my radio component can pass an EDL packet (struct) to the UDP driver.  To simplify this, I'm using an array of U8 instead of the edlpacket_t struct (which requires serialization) to get it working.

Unfortunately, I'm doing something wrong to set this up since it's asserting: 
udpSend_out Assert: " ... fprime/Svc/BufferManager/BufferManagerComponentImpl.cpp:97" 65535 120

I've been trying to debug this with GDB but it's not breaking and stepping properly in that portion of the code.

I would appreciate any help or insight that anyone has to offer. Here is my setup:

topology.fpp:
```
  ...
  topology OreSatDev {
    ...
    instance udpDriver
    instance radio
    ... 
    connections OreSatConnections {
      ...
      radio.udpSend -> udpDriver.$send
      udpDriver.$recv -> radio.udpRecv 
      udpDriver.allocate -> bufferManager.bufferGetCallee
      udpDriver.deallocate -> bufferManager.bufferSendIn
```
instances.fpp:
```
  instance radio: Components.Radio base id 0x1000 \
   queue size Default.QUEUE_SIZE \
   stack size Default.STACK_SIZE \
   priority 93
  ...
  instance bufferManager: Svc.BufferManager base id 0x4400
  ...
  instance udpDriver: Drv.Udp base id 0x4D00
```
OreSatDevTopology.cpp:
```
enum TopologyConstants {
  ...
		EDL_BUFFER_SIZE = 100 * sizeof(U8),
		EDL_BUFFER_COUNT = 30
};
```
...
```
void configureTopology() {
    ...
    upBuffMgrBins.bins[3].bufferSize = EDL_BUFFER_SIZE;
    upBuffMgrBins.bins[3].numBuffers = EDL_BUFFER_COUNT;
    bufferManager.setup(BUFFER_MANAGER_ID, 0, mallocator, upBuffMgrBins);
    ...
}
```
...
```
void setupTopology(const TopologyState& state) {
    ...
		udpDriver.init(0);
		udpDriver.configureSend("127.0.0.1",1331);
		udpDriver.configureRecv("127.0.0.1",1331);
		Os::TaskString name("ReceiveTask");
		udpDriver.start(name, true, COMM_PRIORITY, Default::STACK_SIZE);
}
```
...
```
void teardownTopology(const TopologyState& state) {
    ...
    udpDriver.stop();
    (void)udpDriver.join();
    ...
}
```
Component/Radio/Radio.fpp:
```
  @ UDP send port: sends an EDL packet to the UDP driver
  output port udpSend: Drv.ByteStreamSend
  
  @ UDP recv port: receives an EDL packet from the UDP driver
  guarded input port udpRecv: Drv.ByteStreamRecv
```
Component/Radio/Radio.hpp
```
  //! UDP recv port: receives an EDL packet from the UDP driver
  void udpRecv_handler(
    FwIndexType portNum, //!< The port number
    Fw::Buffer& recvBuffer,
    const Drv::RecvStatus& recvStatus
  ) override;
```
Component/Radio/Radio.cpp:
```
	void Radio :: udpClientTaskEntry(void *ptr) {
		FW_ASSERT(ptr != nullptr);
		Radio *radio = reinterpret_cast<Radio*>(ptr);

		U8 data[80] = {0};
		Fw::Buffer buffer(data, sizeof(data));

		U8 i = 0;
		for (;;++i) {
			data[0] = i;
			printf("udpSend_out ");
			radio->udpSend_out(0, buffer);
			Os::Task::delay(Fw::TimeInterval(1, 0));
		}
	}
```

**Comment by zimri-leisher** [comment, marked as answer]
I think your problem is in Radio.cpp.

You're manually constructing a buffer when you should be getting it from your bufferManager's bufferGetCallee port. So instead of:
```
Fw::Buffer buffer(data, sizeof(data));
```

you should do
```
// allocate a buffer from the bufman
FwBuffer buffer = this->allocate_out(0, sizeof(data));
// check if allocation of sufficient mem failed
if (buffer.getSize() < sizeof(data)) {
    FW_ASSERT(0, buffer.getSize(), sizeof(data));
}
// write the data into the buffer
// args are: U8* arr, len of the arr, and True to stop the serialize method from writing the len of the array into the buf
Fw::SerializeStatus stat = buffer.serialize(data, sizeof(data), True);
FW_ASSERT(stat == Fw::SerializeStatus::FW_SERIALIZE_OK);
```

then you can send the buffer to the driver, which will send it via a socket and return the buffer to the buffer manager.

However, I would note that I'm confunsed why you seem to be declaring a thread entry point:
```
	void Radio :: udpClientTaskEntry(void *ptr) {
```
If you're constructing a separate thread yourself, instead I would recommend using a combination of ActiveRateGroup, LinuxTimer and RateGroupDriver components to run your "udpClientTaskEntry" code on a regular schedule.

**Reply by zimri-leisher** [reply]
Other than that, nice work on figuring out the various drivers and buf mans.
You are rarely supposed to directly construct Fw::Buffers. In this case, you can be sure that it wouldn't work because as soon as that buffer gets passed to udpSend, the pointer it's holding to U8 data[80] is no longer gonna be valid, because data is just a local variable in the function and has no guarantee of being accessible outside of the function.

**Reply by chroco** [reply]
That solved my problem, thank you for your help!

I'm just using the thread to figure out the UDP driver and how to pass buffers. Is there an advantage to using the rate driver and Linux timers over tasks? One of my upcoming projects is to see I can connect the rate driver to a PRU on beaglebone black.

**Reply by zimri-leisher** [reply]
Glad I could help!
There's no reason you can't make your own thread, but using rate group driver+linux timer is the idiomatic way to have some regularly scheduled execution of a code block. That way you get all the nice benefits of the FPrime system--your rate groups will warn you if things go over their cycle time, you have the option of queueing up these cycle calls, you can easily hook up more components such that they get executed on the same thread before or after your scheduled code, etc.

**Reply by chroco** [reply]
That makes sense. Which brings up a question about these queued calls. I changed the UDP client code to use the rate group driver instead of a task. I noticed that the LedBlinker example uses async input port run: Svc.Sched and works happily. However, if I try and use this same async input port in the Radio active component then Radio hangs and eventually asserts from what looks to be a full queue. Other components are still executing normally from what I can tell. If I just use a sync input port on Radio, then everything is happy. How come the Led component can get away with the async port? I don't see how it is handling any queued messages. How do I handle these queues?

**Reply by zimri-leisher** [reply]
So sorry for the late response! Good question.
Active components are called active because they have their own thread implicitly. All that thread does is wait in a loop for things to come on the queue, and when it finds something, it executes it.
If your component is active, adding things to the queue should cause them to be almost immediately executed. If the queue is getting full, then I would profile the runtime performance of the async input ports, to see if they are taking up too much time. This is a common occurrence--if you are putting calls to `portX` on the queue at 25 Hz, but `portX` takes 50 milliseconds to run (10 ms too long), your queue will fill and cause an assert.

**Comment by aliaga17** [comment]
Hi, I am following this thread to implement a similar connection but using a TCP Client in the deployment (very useful this far!). Also a newbie so apologies if the questions below sound too obvious :)

@chroco what part in the FPrime documentation did you follow to reach where you got when you first posted the issue? I do not think I could have figured out the various drivers and buffers through the buffer manager and such.

Some more specific questions:
- I see you declared a "udpRecv_handler(..)" function in the Radio.hpp file, but no implementation in the Radio.cpp file is provided. Similarly, there is a udpClientTaskEntry(...) in Radio.cpp but no declaration in Radio.hpp. Would you mind sharing the implementation for those (or some partial versions?)?
- With respect to @zimri-leisher first comment, did you end up using the udpClientTaskEntry function at all? if not, could you elaborate on what do you and @zimri-leisher mean by "a combination of ActiveRateGroup, LinuxTimer and RateGroupDriver components to run your "udpClientTaskEntry"?

Thank you in advance for your help!

EDIT: There might be one more level of complxity in my project. I am implementing this TCP connection between the OBC and a satellite part that might not always be turned on (in fact, it is off most of the time). How would the TCP Driver handle this? Setting up the tcpClient from the setupTopology function seems to set it up permanently and keep the connection open forever. Is that ok? in other words, is it ok to keep the tcpClient of the OBC listening forever even if the server it is listening to is no operative?

Otherwise, what would be the best way to handle dynamic connection control?

EDIT2: Thinking further about the TCP/UDP connection scenario, specifically about buffers. If buffers are requested using the BufferManager, shouldn't the connection the BufMan ports be directly to the Radio component, and not to the udpDriver? 
From @zimri-leisher answer, it seems like the buffers should be allocated using the buffer manager in Radio.cpp. Then, the radio component would send those buffers to the udpDriver component (maybe then guessing udpDriver still needs to be connected to bufferSendIn to deallocate the received buffer?)
